### ISSUE TITLE
Wrong sequence points generated for if in async void method

### ISSUE THREAD

**Issue by Suchiman** [dotnet/roslyn#12564]
**Version Used**: VS2015.3 / VS2017 RC

**Steps to Reproduce**:

``` csharp
using System;

class Program
{
    static void Main(string[] args)
    {
        Test();
    }

    private static async void Test()
    {
        Exception exception = null;

        if (exception != null)
            Console.WriteLine("Doesn't happen");
    }
}
```
1. Step through with this code
2. `Console.WriteLine` is unreachable by logic
3. Debugger still visually steps over it (without actually executing it though)
4. Remove `async` from `Test`
5. Now it works as expected

**Expected Behavior**:
![2016-07-16_17-02-02](https://cloud.githubusercontent.com/assets/7110884/16895456/248628aa-4b77-11e6-8897-c8e50c26eef1.gif)

**Actual Behavior**:
![2016-07-16_16-57-53](https://cloud.githubusercontent.com/assets/7110884/16895462/6b3c8794-4b77-11e6-955e-33898c24ea05.gif)

**Comment by jaredpar** [comment]
Thanks for reporting.  We will take a look at this.

**Comment by jcouv** [comment]
I ran into this issue as well. 
Tomas showed me some examples of unittests that can be useful to check this scenario (`ForEachStatement_MultiDimensionalArray`).

**Comment by jcouv** [comment]
Taking some notes in case I have to table this investigation for a while: the generated IL includes a `leave` right after the instructions for the conditional. If the condition is false, the execution jumps to that `leave` and the debugger pauses, but the only sequence point that precedes it is in the statement/body for the `if`.

From discussion with Tomas and Vlad, the solution seems to be adding a hidden sequence point on the `afterif:` label during lowering for the if statement. 
But from my verification so far, the problem still repro'ed after this fix. So further investigation is needed.
My work so far is stashed at https://github.com/jcouv/roslyn/tree/missing-seq-points

**Comment by jcouv** [comment]
Fixed by #16803 (Missing sequence point on implicit return in async method)
